frontend/src/main/scala/bloop/engine/BuildLoader.scala

```diff
@@ -90,6 +105,89 @@ object BuildLoader {
     logger.debug(s"Loading ${configFiles.length} projects from '${configDir.syntax}'...")(
       DebugFilter.Compilation
     )
-    configFiles.map(f => Project.fromBytesAndOrigin(f.bytes, f.origin, logger))
+    LoadedBuild(configFiles.map(f => loadProject(f.bytes, f.origin, logger, settings)), settings)
+  }
+
+  private def loadProject(
+      bytes: Array[Byte],
+      origin: Origin,
+      logger: Logger,
+      settings: Option[WorkspaceSettings]
+  ): Project = {
+    val project = Project.fromBytesAndOrigin(bytes, origin, logger)
+    settings.map(applySettings(_, project, logger)).getOrElse(project)
+  }
+
+  private def updateWorkspaceSettings(
+      configDir: AbsolutePath,
+      logger: Logger,
+      incomingSettings: Option[WorkspaceSettings]
+  ): Option[WorkspaceSettings] = {
+    val savedSettings = WorkspaceSettings.fromFile(configDir, logger)
+    incomingSettings match {
+      case Some(incoming) =>
+        if (savedSettings.isEmpty || savedSettings.exists(_ != incoming)) {
+          WorkspaceSettings.write(configDir, incoming)
+          Some(incoming)
+        } else {
+          savedSettings
+        }
+      case None =>
+        savedSettings
+    }
+
   }
+
+  /**
+   * Applies workspace settings from bloop.settings.json file to a project. This includes:
+   * - SemanticDB plugin version to resolve and include in Scala compiler options
+   */
+  private def applySettings(
+      settings: WorkspaceSettings,
+      project: Project,
+      logger: Logger
+  ): Project = {
+    def addSemanticDBOptions(pluginPath: AbsolutePath) = {
+      {
+        val optionsSet = project.scalacOptions
+        val containsSemanticDB = optionsSet.find(
+          setting => setting.contains("-Xplugin") && setting.contains("semanticdb-scalac")
+        )
+        val containsYrangepos = optionsSet.find(_.contains("-Yrangepos"))
+        val semanticDBAdded = if (containsSemanticDB.isDefined) {
+          logger.info(s"SemanticDB plugin already added: ${containsSemanticDB.get}")
+          optionsSet
+        } else {
+          val workspaceDir = project.origin.path.getParent.getParent
```

Comment:
@jvican Not sure if that is the best way of getting the workspace root directory.